casper/src/main/resources/PoS.rhox

```diff
@@ -0,0 +1,164 @@
+----+------------+--------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------
+*/
+
+new PoS,
+    registryLookup(`rho:registry:lookup`),
+    revAddressOps(`rho:rev:address`),
+    revVaultCh,
+    getUser,
+    getCurrentUserAddress,
+    getCurrentUserVault,
+    getMVar,
+    runMVar,
+    deposit in {
+  registryLookup!(`rho:id:1o93uitkrjfubh43jt19owanuezhntag5wh74c6ur5feuotpi73q8z`, *revVaultCh) |
+
+  for(@(_, RevVault) <- revVaultCh) {
+    new posRevAddressCh in {
+
+      revAddressOps!("fromPublicKey", "cc87bf7747a8c176714b417ca14a63897b07572876c5e38a7896b6007738ef81".hexToBytes(), *posRevAddressCh) |
+
+      for(@posRevAddress <- posRevAddressCh) {
+        new bondsCh in {
+          bondsCh!({}) |
+          getCurrentUserVault!(Nil) |
+
+          contract PoS (@"getBonds", returnCh) = {
+            getMVar!(*bondsCh, *returnCh)
+          } |
+
+          contract PoS (@"bond", @amount, returnCh) = {
+            new userCh, depositCh, processCh in {
+              runMVar!(*bondsCh, *processCh, *returnCh) |
+
+              getUser!(*userCh) |
+              for(@bonds, resultCh <- processCh;
+                  @userPk <- userCh) {
+                if (bonds.contains(userPk)) {
+                  resultCh!(bonds, (false, "Public key is already bonded."))
+                } else if (amount < $$minimumBond$$) {
+                  resultCh!(bonds, (false, "Bond is less than minimum!"))
+                } else if (amount > $$maximumBond$$) {
+                  resultCh!(bonds, (false, "Bond is greater than maximum!"))
+                } else {
+                  deposit!(userPk, amount, *depositCh) |
+                  for(@depositResult <- depositCh) {
+                    match depositResult {
+                      (true, _) => {
+                        resultCh!(
+                          bonds.set(userPk,[amount, bonds.size() + 1]),
+                          depositResult)
+                      }
+
+                      (false, errorMsg) => {
+                        resultCh!(bonds, (false, "Bond deposit failed: " ++ errorMsg))
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          } |
+
+          contract PoS (@"pay", @amount, returnCh) = {
+            new vaultCh, transferCh, userCh in {
+              getUser!(*userCh) |
+
+              for(@userPk <- userCh) {
+                deposit!(userPk, amount, *returnCh)
+              }
+            }
+          }
+        } |
+
+        contract deposit (@fromUserPk, @amount, returnCh) = {
+          new vaultCh,
+              revAddressCh,
+              authKeyCh in {
+            @RevVault!("deployerAuthKey", *authKeyCh) |
+            for (@authKey <- authKeyCh) {
+              revAddressOps!("fromPublicKey", fromUserPk, *revAddressCh) |
```

Comment:
this relies direclty on the 'drain vault attack' that is going to be prevented. Instead of doing the deposit for the bonded party, we should let them do that, notify us when it's done, check balance, and proceed accordingly.

This also means each bonding request should receive a new, one-time-only, unforgrable-name-bound RevVault to which the bonding amount would be transferred by the party, and from it to the PoS vault.

Reply:
This complicates the API. I wouldn't worry about the PoS contract receiving a rootAuth token when needed because it's a "system contract"

Let's fix the vault drain issue first and then we will see how this changes.

Reply:
Please create a new issue for the one-time-stake-transfer rev vaults and put it into sprint-30 :)

Reply:
I'm not sure we actually need that one-time-stake-transfer. If we consider acceptable to have the rootAuth passed to the RevVault contract then we can also have it passed to other system contracts like the PoS contract.

Reply:
The one time stake transfer is needed, because two bonding requests could be processed in parallel, leading to the balance not matching what we expect.

`If we consider acceptable to have the rootAuth passed to the RevVault contract ` - this is only for creating a RevVault-specific auth ocap. After having on-chain-deployer-id we could just as well remove the `deployerAuthKey` from RevVault and make the user create that token themselves based on a `bundle0{_revVault}` exposed by the RevVault contract. So they'd call
`@AuthKey!("make", (*_revVaultOcapBundled, bundle0{*deployerId}), *ret)` to create a key themselves. I'm not sure why you bring this up in this thread though.

Reply:
Ah, I see why you bring that up. In no case should passing a `deployerId` to a contract should be utilized to do things in other contracts. If the user wants to give us that capability, they should provide the related tokens.

Reply:
How are you going to prevent that? The deployerId is, in fact, a `su - root`. With this in mind we should be aware of the risks involved and we could take advantage of it, one of the situations being this one.

Anyway, we should discuss this in detail: https://rchain.atlassian.net/browse/RCHAIN-3437